### ISSUE TITLE
what happen if row is full and all keys  are in same logic table

### ISSUE THREAD

**Issue by Masterlvng** [simonhf/sharedhashfile#3]
Add 16 keys into shf.  It's possible that all the keys locate in same row and with same logic table.
In this situation, split is useless. if i add another key, still locate in same row and with same logic table, it will fail.

**Comment by simonhf** [comment]
Thanks for the comment. I don't see this happening in a real world scenario, and shf has been tested with many hundreds of millions of unique keys in a single table. Please note that when split happens then more hashed key bits get used. If you think otherwise then please create a failing  test illustrating the problem and I will gladly fix it.

**Comment by Masterlvng** [comment]
I think,  if deploy it as a long run system,  the situation may happen, although the  possibility is relatively small. if i say wrong, please to correct me with example.
If happen, it 'll make me confused. maybe the whole shf is not full, but i fail to add some keys.
I think this idea is imcomplete.

BTW, when i first saw this project, i m really excited. Your code is really nice, and the design of 'split' is really attracting me.

**Comment by simonhf** [comment]
Okay, here's a quick program which may or may not convince you :-) shf can store up to 2^32 keys if enough RAM is available. A 32 bit hash is used to locate each key. For the recursive split to fail then it'd be necessary to have 16 keys which all have the same 32 bit hash. The program below hashes a billion keys -- probably the most you would ever store in shf due to physical RAM limitations -- and the highest number of duplicated 32 bit key hashes is 8. Even if you had a long running system with 1 billion keys stored, and those keys constantly changing, then there's no reason to assume that the randomness of the hashing would become worse for a certain set of 1 billion keys, or?

In theory a malicious user who understands shf source code could pre-calculate 17 keys which hash to the same 32-bit hash and try to inject those keys into the system somehow. Therefore, a good update to shf would be to randomize the seed so that such an attack cannot happen.

Hope this helps!

```
$ gcc -O3 -o hash-dups hash-dups.c && ./hash-dups 
- allocating 4294967296 bytes
- key 594,824,057 has hash 0xdea21c2a which occurs 7 times
- key 686,006,630 has hash 0xa3f184c7 which occurs 7 times
- key 732,702,904 has hash 0x9c446fa9 which occurs 7 times
- key 824,279,625 has hash 0xdbd6229d which occurs 7 times
- key 824,713,304 has hash 0x2ed295af which occurs 7 times
- key 842,041,876 has hash 0x8c8203f9 which occurs 7 times
- key 847,552,345 has hash 0x16fc92ae which occurs 7 times
- key 873,712,235 has hash 0x16fc92ae which occurs 8 times
- key 877,700,588 has hash 0x2d7b480d which occurs 7 times
- key 904,074,112 has hash 0xed1192e9 which occurs 7 times
- key 909,651,712 has hash 0x451efc1a which occurs 7 times
- key 911,979,816 has hash 0x019b86d2 which occurs 7 times
- key 938,567,927 has hash 0xba09d4b0 which occurs 7 times
- key 944,499,264 has hash 0x939ca797 which occurs 7 times
- key 949,948,690 has hash 0x2b941c8c which occurs 7 times
- key 951,883,864 has hash 0x15c6151e which occurs 7 times
- key 953,808,006 has hash 0x538d43d8 which occurs 7 times
- key 957,081,746 has hash 0x5fa165c9 which occurs 7 times
- key 962,485,221 has hash 0x6a44c890 which occurs 7 times
- key 965,732,568 has hash 0x615afd81 which occurs 7 times
- key 967,292,754 has hash 0x7e210044 which occurs 7 times
- key 980,081,444 has hash 0xc2800ff2 which occurs 7 times
- key 983,687,510 has hash 0x4a50cb65 which occurs 7 times
- key 989,469,524 has hash 0x54c675fc which occurs 7 times
- key 999,153,280 has hash 0x14f509cb which occurs 7 times
```

```
#include <stdio.h>
#include <sys/mman.h>
#include <locale.h>
#include "shf.private.h"
#include "murmurhash3.c"

void main(void)
{
    setlocale(LC_NUMERIC, ""); // for printf commas
    size_t table_size = 1;
    printf("- allocating %lu bytes\n", table_size << 32);
    uint8_t * table = mmap (NULL, table_size << 32, PROT_READ|PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_NORESERVE, -1, (off_t)0);
    if (MAP_FAILED == table) {
        printf("mmap() failed\n");
        exit(1);
    }
    uint64_t key = 0;
    SHF_HASH shf_hash;
    while (key < 1000000000) {
        MurmurHash3_x64_128 ((const void *)&key, sizeof(key), 1967, &shf_hash.u64[0]);
        //printf("- key is %lu, hash is 0x%08x\n", key, shf_hash.u32[0]);
        table[shf_hash.u32[0]] ++;
        if (table[shf_hash.u32[0]] >= 7) {
            printf("- key %'lu has hash 0x%08x which occurs %u times\n", key, shf_hash.u32[0], table[shf_hash.u32[0]]);
        }
        key ++;
    }
}
```

**Comment by Masterlvng** [comment]
Really thanks for you jobs. This result make me convinced.
The project is awesome. Do you have any plan to continue developing?

**Comment by Masterlvng** [comment]
I think the following program is more reasonable.
I dont have enough RAM, can you please run it?
```c
#include <stdio.h>
#include <sys/mman.h>
#include <locale.h>
#include "shf.private.h"
#include "murmurhash3.c"

void main(void)
{
    setlocale(LC_NUMERIC, ""); // for printf commas
    size_t table_size = 1;
    printf("- allocating %lu bytes\n", table_size << 32);
    uint8_t * table = mmap (NULL, table_size << 32, PROT_READ|PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_NORESERVE, -1, (off_t)0);
    if (MAP_FAILED == table) {
        printf("mmap() failed\n");
        exit(1);
    }
    uint64_t key = 0;
    SHF_HASH shf_hash;
    while (key < 1000000000) {
        MurmurHash3_x64_128 ((const void *)&key, sizeof(key), 1967, &shf_hash.u64[0]);
        //printf("- key is %lu, hash is 0x%08x\n", key, shf_hash.u32[0]);
        uint32_t win    = shf_hash.u16[0] %             SHF_WINS_PER_SHF       ;
        uint32_t tab2  = shf_hash.u16[1] %             SHF_TABS_PER_WIN       ;
        uint32_t row   = shf_hash.u16[2] %             SHF_ROWS_PER_TAB      ;
        SHF_UID uid;
        uid.as_part.win = win;
        uid.as_part.tab = tab2;
        uid.as_part.row = row;
        table[uid.as_u32] ++;
        if (table[uid.as_u32] >= 7) {
            printf("- key %'lu has hash 0x%08x which occurs %u times\n", key, uid.as_u32, table[uid.as_u32]);
        }
        key ++;
    }
}
```

**Comment by Masterlvng** [comment]
i ran the program above. only print key which occurs more than 15 times
some result 
```
- key 895418353 has hash 0x6a05cce7 which occurs 17 times
- key 895494772 has hash 0x62da6e2d which occurs 16 times
- key 895593610 has hash 0x602fe7f3 which occurs 16 times
- key 895784918 has hash 0x669a0604 which occurs 16 times
- key 895992127 has hash 0x6cf98baf which occurs 16 times
- key 896234367 has hash 0x6bdff86d which occurs 16 times
- key 897340689 has hash 0x66697cdc which occurs 16 times
- key 897957535 has hash 0x697df2e7 which occurs 16 times
- key 897962340 has hash 0x6796b3bf which occurs 16 times
- key 898417517 has hash 0x6ef1ed1c which occurs 16 times
- key 898452064 has hash 0x676d0992 which occurs 16 times
- key 899206380 has hash 0x63f79831 which occurs 17 times
- key 899532495 has hash 0x67dd9f79 which occurs 16 times
- key 899809450 has hash 0x62738be7 which occurs 16 times
- key 899826040 has hash 0x6c9bd728 which occurs 17 times
- key 900531309 has hash 0x6a232be5 which occurs 16 times 
```
the result make me confused again : (

**Comment by simonhf** [comment]
Thanks for modifying the program and looking at this further. I agree with your results. My bad, the algorithm only uses 28 bits of the key hash and not 32 bits; only the UID is 32 bits. This means the over flow situation can occur above 500 million keys, so much lower than previously thought.

Regarding a workaround for boxes which have enough RAM to store over 500 million key,value pairs: I suggest having multiple shfs on the same box and sharding between them, e.g. if you want to grow to 8 billion keys then have at least 16 shf instances in parallel. And sharding should not add much overhead to key operations because key hashing is already a separate operation and an unused part of the key hash can be used to shard.

The only down side of the workaround is that the 32 bit UID can no longer be used to reference a unique key,value pair. Instead, you'd need the 32 bit UID and e.g. an 8 bit shard value. Using an 8 bit shard value -- so up to 256 shf instances -- then the upper limit on the number of keys is about 128 billion keys; 500 million * 256.

Hope this helps.

**Comment by Masterlvng** [comment]
Thanks for your reply and suggestion.
In my opinion, this design may be increase the burden and risks to users.
I keep thinking how to solve this problem, but no good idea so far.
Without this problem, the project will be more powerful : )

**Comment by simonhf** [comment]
Thanks for the message and please let me know if you come up with any other ideas.

Without the sharding then the code looks like this:

```
    shf_make_hash(SHF_CONST_STR_AND_SIZE("key"));
    uint32_t uid = shf_put_key_val(shf, "val", 3);
```

And with the sharding the code looks like this:

```
    shf_make_hash(SHF_CONST_STR_AND_SIZE("key"));
    uint8_t  shard = shf_hash.u08[6];
    uint32_t uid = shf_put_key_val(shf[shard], "val", 3);
```

Which burdens and/or risks do you see?

**Comment by Masterlvng** [comment]
I mean, a storage system should not tell users "if your keys are more than 500 million keys, it will appear some problems. Let you fail to insert some keys. if you want to aviod such problem, you should shard your data".  This statement  increases burdens to users.

I will try to figure out a solution ：)

BTW,  Is there any case,which use shf in production?

**Comment by simonhf** [comment]
hehe... but I think that any user who is going to store more than 500 million keys (a) probably knows what they are doing, and (b) will probably not be able to use any other key value store because most of the others are slower, take more storage, or don't offer special functionality, or all of the above.

Another solution would be to have a single shf instance but implement a 64 bit UID, which would definitely work with the splitting algorithm. I also avoided this during the design because I wanted to keep the UID as small as possible. But I'm looking forward to hearing any ideas you come up with :-)

shf is used in production but AFAIK not yet with extreme numbers of keys but I don't know for sure.

**Comment by simonhf** [comment]
Here's an idea albeit one with caveats:

Once no more splitting is possible and each tab is 'unique', and a full row has been detected inside the tab -- i.e. the overflow condition -- then an extra bit could be set in the tab index to signify this condition:

```
typedef struct SHF_OFF_MMAP {
    volatile uint16_t tab; /* index up to 2048 tabs; todo use extra bit for something :-) */
} __attribute__((packed)) SHF_OFF_MMAP;
```

This extra bit would mean 'use extra rows when accessing a key'. So instead of the key having to be in a particular row in one of 16 different positions, it could as well be in different rows in the tab.

The disadvantage of this mechanism is that accessing a key,value pair using the key (in a tab in the overflow condition) will get slower because more searching around must be done. However, it maybe possible to minimize the hit by doing the search in parallel using the CPU prefetch instruction.

However, the advantage of this mechanism is that the UID mechanism remains the same and just as fast. After storing a key,value in an 'overflow row' then the 32 bit UID returned can still point directly to the row used.

So in summary: 'overflow tabs' would cause access by key to get slower, whereas access by UID would remain just as fast... What do you think?

**Comment by Masterlvng** [comment]
Sorry for late reply. 
```
it could as well be in different rows in the tab
```
So, in order to access by key,  you must keep some information that indicating next row to search if the current row is full.  Am I right?

**Comment by simonhf** [comment]
Yes, the next row to search could be determined by information or by algorithm of which there are lots here [1].

[1] https://en.wikipedia.org/wiki/Hash_table#Collision_resolution

**Comment by Masterlvng** [comment]
I think shf should add more logic when shrinking, splitting etc.
Look forward to see the implement. : )
